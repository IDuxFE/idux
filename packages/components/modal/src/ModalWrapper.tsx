/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import {
  type ComputedRef,
  type Ref,
  Transition,
  computed,
  defineComponent,
  inject,
  onBeforeUnmount,
  onMounted,
  ref,
  watch,
} from 'vue'

import { isFunction } from 'lodash-es'

import { useDraggable } from '@idux/cdk/drag-drop'
import { callEmit, convertCssPixel, getOffset } from '@idux/cdk/utils'
import { ɵFooter } from '@idux/components/_private/footer'
import { ɵHeader } from '@idux/components/_private/header'
import { type ModalConfig } from '@idux/components/config'

import ModalBody from './ModalBody'
import { MODAL_TOKEN, modalToken } from './token'
import { type ModalProps } from './types'

export default defineComponent({
  inheritAttrs: false,
  setup(_, { attrs }) {
    const {
      props,
      slots,
      common,
      locale,
      config,
      mergedPrefixCls,
      visible,
      animatedVisible,
      mergedVisible,
      cancelLoading,
      okLoading,
      currentZIndex,
    } = inject(modalToken)!
    const { close, cancel, ok } = inject(MODAL_TOKEN)!
    const { centered, closable, closeIcon, closeOnEsc, width, mask, maskClosable } = useConfig(props, config)

    const cancelVisible = computed(() => props.type === 'default' || props.type === 'confirm')

    const cancelText = computed(() => props.cancelText ?? locale.modal.cancelText)
    const okText = computed(() => {
      if (props.okText) {
        return props.okText
      }
      return cancelVisible.value ? locale.modal.okText : locale.modal.justOkText
    })

    const placementStyle = computed(() => {
      const top = centered.value ? 0 : convertCssPixel(props.offset)
      return { top, width: width.value }
    })

    const wrapperClasses = computed(() => {
      const { wrapperClassName = '' } = props
      const prefixCls = mergedPrefixCls.value

      return {
        [`${prefixCls}-wrapper`]: true,
        [`${prefixCls}-centered`]: centered.value,
        [`${prefixCls}-with-mask`]: mask.value,
        [wrapperClassName]: !!wrapperClassName,
      }
    })

    const wrapperStyle = computed(() => {
      return { zIndex: currentZIndex.value }
    })

    const modalTransformOrigin = ref<string>()
    const contentStyle = computed(() => {
      return { transformOrigin: modalTransformOrigin.value, ...placementStyle.value }
    })

    const wrapperRef = ref<HTMLDivElement>()
    const modalRef = ref<HTMLDivElement>()
    const contentRef = ref<HTMLDivElement>()
    const headerRef = ref<HTMLDivElement>()
    const sentinelStartRef = ref<HTMLDivElement>()
    const sentinelEndRef = ref<HTMLDivElement>()

    const { onWrapperClick, onWrapperKeydown, onContentMousedown, onContentMouseup } = useEvent(
      props,
      close,
      mask,
      maskClosable,
      closeOnEsc,
      sentinelStartRef,
      sentinelEndRef,
    )

    const { onEnter, onAfterEnter, onAfterLeave } = useEvents(
      props,
      wrapperRef,
      modalRef,
      animatedVisible,
      modalTransformOrigin,
    )
    const draggableResult = ref()

    const handleOk = (evt?: unknown) => {
      if (visible.value) {
        ok(evt)
      }
    }
    const handleCancel = (evt?: unknown) => {
      if (visible.value) {
        cancel(evt)
      }
    }

    onMounted(() => watchVisibleChange(props, wrapperRef, sentinelStartRef, mask, draggableResult))

    watch(
      () => props.draggable,
      draggable => {
        if (draggableResult.value) {
          draggableResult.value.stop()
          draggableResult.value = undefined
        }
        if (draggable) {
          draggableResult.value = useDraggable(contentRef, { handle: headerRef, free: true, boundary: window })
        }
      },
      { immediate: true },
    )

    return () => {
      const prefixCls = mergedPrefixCls.value

      return (
        <div
          v-show={mergedVisible.value}
          ref={wrapperRef}
          class={wrapperClasses.value}
          style={wrapperStyle.value}
          tabindex={-1}
          onClick={onWrapperClick}
          onKeydown={onWrapperKeydown}
        >
          <Transition
            name={props.animatable ? `${common.prefixCls}-zoom` : undefined}
            appear
            onEnter={onEnter}
            onAfterEnter={onAfterEnter}
            onAfterLeave={onAfterLeave}
          >
            <div
              v-show={visible.value}
              ref={modalRef}
              role="document"
              class={prefixCls}
              style={contentStyle.value}
              onMousedown={onContentMousedown}
              onMouseup={onContentMouseup}
              {...attrs}
            >
              <div ref={sentinelStartRef} tabindex={0} class={`${prefixCls}-sentinel`} aria-hidden={true}></div>
              <div ref={contentRef} class={`${prefixCls}-content`}>
                <ɵHeader
                  ref={headerRef}
                  v-slots={slots}
                  closable={closable.value}
                  closeIcon={closeIcon.value}
                  header={props.header}
                  onClose={close}
                />
                <ModalBody></ModalBody>
                <ɵFooter
                  v-slots={slots}
                  class={`${prefixCls}-footer`}
                  cancel={handleCancel}
                  cancelButton={props.cancelButton}
                  cancelLoading={cancelLoading.value}
                  cancelText={cancelText.value}
                  cancelVisible={cancelVisible.value}
                  footer={props.footer}
                  ok={handleOk}
                  okButton={props.okButton}
                  okLoading={okLoading.value}
                  okText={okText.value}
                ></ɵFooter>
              </div>
              <div ref={sentinelEndRef} tabindex={0} class={`${prefixCls}-sentinel`} aria-hidden={true}></div>
            </div>
          </Transition>
        </div>
      )
    }
  },
})

function useConfig(props: ModalProps, config: ModalConfig) {
  const centered = computed(() => props.centered ?? config.centered)
  const closable = computed(() => props.closable ?? config.closable)
  const closeIcon = computed(() => props.closeIcon ?? config.closeIcon)
  const closeOnEsc = computed(() => props.closeOnEsc ?? config.closeOnEsc)
  const mask = computed(() => props.mask ?? config.mask)
  const maskClosable = computed(() => props.maskClosable ?? config.maskClosable)
  const width = computed(() => convertCssPixel(props.width ?? config.width))

  return { centered, closable, closeIcon, closeOnEsc, width, mask, maskClosable }
}

function watchVisibleChange(
  props: ModalProps,
  wrapperRef: Ref<HTMLDivElement | undefined>,
  sentinelStartRef: Ref<HTMLDivElement | undefined>,
  mask: ComputedRef<boolean>,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  draggableResult: Ref<any>,
) {
  let lastOutSideActiveElement: HTMLElement | null = null

  watch(
    () => props.visible,
    visible => {
      if (visible) {
        const wrapperElement = wrapperRef.value!
        const activeElement = document.activeElement
        if (!wrapperElement.contains(activeElement)) {
          lastOutSideActiveElement = activeElement as HTMLElement
          sentinelStartRef.value?.focus()
        }
        draggableResult.value?.reset()
      } else {
        if (mask.value) {
          lastOutSideActiveElement?.focus?.()
          lastOutSideActiveElement = null
        }
      }
    },
    { immediate: true },
  )
}

function useEvent(
  props: ModalProps,
  close: (evt: Event) => void,
  mask: ComputedRef<boolean>,
  maskClosable: ComputedRef<boolean>,
  closeOnEsc: ComputedRef<boolean>,
  sentinelStartRef: Ref<HTMLDivElement | undefined>,
  sentinelEndRef: Ref<HTMLDivElement | undefined>,
) {
  let timeId: number | undefined
  let mouseDown = false

  const clearTimer = () => {
    if (timeId) {
      clearTimeout(timeId)
      timeId = undefined
    }
  }

  const onWrapperClick = (evt: MouseEvent) => {
    if (evt.target === evt.currentTarget && (!mouseDown || props.draggable) && mask.value && maskClosable.value) {
      close(evt)
    }
  }

  const onWrapperKeydown = (evt: KeyboardEvent) => {
    if (closeOnEsc.value && evt.code === 'Escape') {
      evt.stopPropagation()
      close(evt)
      return
    }

    if (evt.code === 'Tab') {
      const activeElement = document.activeElement
      const sentinelStartElement = sentinelStartRef.value
      const sentinelEndElement = sentinelEndRef.value
      if (evt.shiftKey) {
        if (activeElement === sentinelStartElement) {
          sentinelEndElement?.focus()
        }
      } else if (activeElement === sentinelEndElement) {
        sentinelStartElement?.focus()
      }
    }
  }

  const onContentMousedown = () => {
    clearTimer()
    mouseDown = true
  }

  const onContentMouseup = () => {
    if (mouseDown) {
      timeId = setTimeout(() => (mouseDown = false))
    }
  }

  onBeforeUnmount(() => clearTimer())

  return { onWrapperClick, onWrapperKeydown, onContentMousedown, onContentMouseup }
}

function useEvents(
  props: ModalProps,
  wrapperRef: Ref<HTMLDivElement | undefined>,
  modalRef: Ref<HTMLDivElement | undefined>,
  animatedVisible: Ref<boolean | undefined>,
  modalTransformOrigin: Ref<string | undefined>,
) {
  let lastOutSideActiveElement: HTMLElement | null = null
  const onEnter = () => {
    const wrapperElement = wrapperRef.value!
    const activeElement = document.activeElement
    if (!wrapperElement.contains(activeElement)) {
      lastOutSideActiveElement = activeElement as HTMLElement
    }

    // first show modal
    if (!modalTransformOrigin.value && lastOutSideActiveElement) {
      const modalElement = modalRef.value!
      const previouslyDOMRect = lastOutSideActiveElement.getBoundingClientRect()
      const lastPosition = getOffset(lastOutSideActiveElement)
      const x = lastPosition.left + previouslyDOMRect.width / 2
      const y = lastPosition.top + previouslyDOMRect.height / 2
      modalTransformOrigin.value = `${x - modalElement.offsetLeft}px ${y - modalElement.offsetTop}px`
    }
  }

  const onAfterEnter = () => {
    const wrapperElement = wrapperRef.value!
    const activeElement = document.activeElement
    if (!wrapperElement.contains(activeElement)) {
      wrapperElement.focus()
    }

    callEmit(props.onAfterOpen)
    animatedVisible.value = true
  }

  const onAfterLeave = () => {
    if (lastOutSideActiveElement && isFunction(lastOutSideActiveElement.focus)) {
      const wrapperElement = wrapperRef.value!
      const activeElement = document.activeElement

      if (
        !activeElement ||
        activeElement === document.body ||
        activeElement === wrapperElement ||
        wrapperElement.contains(activeElement)
      ) {
        lastOutSideActiveElement.focus()
      }
    }

    callEmit(props.onAfterClose)
    animatedVisible.value = false
  }
  return { onEnter, onAfterEnter, onAfterLeave }
}
