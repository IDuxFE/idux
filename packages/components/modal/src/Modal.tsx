/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { ModalProps } from './types'
import type { ComputedRef } from 'vue'

import { computed, defineComponent, onBeforeUnmount, provide, ref, watchEffect } from 'vue'

import { CdkPortal } from '@idux/cdk/portal'
import { BlockScrollStrategy } from '@idux/cdk/scroll'
import { callEmit, isPromise, useControlledProp } from '@idux/cdk/utils'
import { ɵMask } from '@idux/components/_private'
import { useGlobalConfig } from '@idux/components/config'

import ModalWrapper from './ModalWrapper'
import { MODAL_TOKEN, modalToken } from './token'
import { modalProps } from './types'

export default defineComponent({
  name: 'IxModal',
  inheritAttrs: false,
  props: modalProps,
  setup(props, { slots, expose, attrs }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-modal`)
    const config = useGlobalConfig('modal')
    const mask = computed(() => props.mask ?? config.mask)
    const zIndex = computed(() => props.zIndex ?? config.zIndex)
    const { visible, setVisible, animatedVisible, mergedVisible } = useVisible(props, mask)
    const { cancelLoading, okLoading, open, close, cancel, ok } = useTrigger(props, setVisible)

    provide(modalToken, {
      props,
      slots,
      common,
      config,
      mergedPrefixCls,
      visible,
      animatedVisible,
      mergedVisible,
      cancelLoading,
      okLoading,
    })

    const apis = { open, close, cancel, ok }
    provide(MODAL_TOKEN, apis)
    expose(apis)

    return () => {
      if (!mergedVisible.value && props.destroyOnHide) {
        return null
      }

      return (
        <CdkPortal target={`${mergedPrefixCls.value}-container`} load={visible.value}>
          <ɵMask
            class={`${mergedPrefixCls.value}-mask`}
            mask={mask.value}
            visible={visible.value}
            zIndex={zIndex.value}
          />
          <ModalWrapper {...attrs}></ModalWrapper>
        </CdkPortal>
      )
    }
  },
})

function useVisible(props: ModalProps, mask: ComputedRef<boolean>) {
  const [visible, setVisible] = useControlledProp(props, 'visible', false)

  const animatedVisible = ref<boolean>()

  const mergedVisible = computed(() => {
    const currVisible = visible.value
    const currAnimatedVisible = animatedVisible.value
    if (currAnimatedVisible === undefined || currVisible) {
      return currVisible
    }
    return currAnimatedVisible
  })

  let scrollStrategy: BlockScrollStrategy | undefined

  watchEffect(() => {
    if (mask.value) {
      if (mergedVisible.value) {
        if (!scrollStrategy) {
          scrollStrategy = new BlockScrollStrategy()
        }
        scrollStrategy.enable()
      } else {
        scrollStrategy?.disable()
      }
    }
  })

  onBeforeUnmount(() => scrollStrategy?.disable())

  return { visible, setVisible, animatedVisible, mergedVisible }
}

function useTrigger(props: ModalProps, setVisible: (value: boolean) => void) {
  const open = () => setVisible(true)

  const close = async (evt?: Event | unknown) => {
    const result = await callEmit(props.onClose, evt)
    if (result === false) {
      return
    }
    setVisible(false)
  }

  const cancelLoading = ref(false)
  const cancel = async (evt?: Event | unknown) => {
    let result = callEmit(props.onCancel, evt)
    if (isPromise(result)) {
      cancelLoading.value = true
      result = await result
      cancelLoading.value = false
    }
    if (result === false) {
      return
    }
    setVisible(false)
  }

  const okLoading = ref(false)
  const ok = async (evt?: Event | unknown) => {
    let result = callEmit(props.onOk, evt)
    if (isPromise(result)) {
      okLoading.value = true
      result = await result
      okLoading.value = false
    }
    if (result === false) {
      return
    }
    setVisible(false)
  }

  return { cancelLoading, okLoading, open, close, cancel, ok }
}
