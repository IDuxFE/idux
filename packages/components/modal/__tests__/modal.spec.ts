import { MountingOptions, VueWrapper, flushPromises, mount } from '@vue/test-utils'
import { VNode, createTextVNode, h } from 'vue'

import { isElementVisible, renderWork } from '@tests'

import { IxButton } from '@idux/components/button'
import { HeaderProps, IxHeader } from '@idux/components/header'
import { IxIcon } from '@idux/components/icon'

import Modal from '../src/Modal'
import ModalWrapper from '../src/ModalWrapper'
import { ModalButtonProps, ModalInstance, ModalProps } from '../src/types'

describe('Modal', () => {
  const content = [h('p', 'Some contents...')]
  const ModalMount = (options?: MountingOptions<Partial<ModalProps>>) => {
    const { props, slots, ...rest } = options || {}
    const _options = {
      props: { visible: true, ...props },
      slots: { default: content, ...slots },
      attachTo: 'body',
      ...rest,
    } as MountingOptions<ModalProps>
    return mount(Modal, _options) as VueWrapper<ModalInstance>
  }

  afterEach(() => {
    document.querySelector('.ix-modal-container')!.innerHTML = ''
  })

  renderWork<ModalProps>(Modal, {
    props: { visible: true },
    slots: { default: () => content },
    attachTo: 'body',
  })

  test('v-model:visible work', async () => {
    const onUpdateVisible = vi.fn()
    const wrapper = ModalMount({ props: { visible: false, 'onUpdate:visible': onUpdateVisible } })
    expect(isElementVisible(document.querySelector('.ix-modal-wrapper'))).toBe(false)

    await wrapper.setProps({ visible: true })

    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal').isVisible()).toBe(true)

    await modalWrapper.find('.ix-button').trigger('click')

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('custom button work', async () => {
    let cancelText = 'No'
    const cancelButton = { mode: 'dashed' } as const
    let okText = 'Yes'
    const okButton = { mode: 'primary', danger: true } as const

    const wrapper = ModalMount({ props: { cancelText, cancelButton, okText, okButton } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    const [okButtonWrapper, cancelButtonWrapper] = await modalWrapper.findAll('.ix-button')

    expect(cancelButtonWrapper.text()).toBe(cancelText)
    expect(cancelButtonWrapper.classes()).toContain('ix-button-dashed')
    expect(okButtonWrapper.text()).toBe(okText)
    expect(okButtonWrapper.classes()).toContain('ix-button-primary')
    expect(okButtonWrapper.classes()).toContain('ix-button-danger')

    cancelText = 'No No'
    okText = 'Yes Yes'
    await wrapper.setProps({ cancelText, okText })

    expect(cancelButtonWrapper.text()).toBe(cancelText)
    expect(okButtonWrapper.text()).toBe(okText)
  })

  test('centered work', async () => {
    const wrapper = ModalMount({ props: { centered: true } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.classes()).toContain('ix-modal-centered')

    await wrapper.setProps({ centered: false })

    expect(modalWrapper.classes()).not.toContain('ix-modal-centered')
  })

  test('closable work', async () => {
    const wrapper = ModalMount({ props: { closable: false } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-icon-close').exists()).toBe(false)

    await wrapper.setProps({ closable: true })

    expect(modalWrapper.find('.ix-icon-close').exists()).toBe(true)
  })

  test('closeIcon work', async () => {
    const onClose = vi.fn()
    const wrapper = ModalMount({ props: { closeIcon: 'up', onClose } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-icon-close').exists()).toBe(false)
    expect(modalWrapper.find('.ix-icon-up').exists()).toBe(true)

    await modalWrapper.find('.ix-icon-up').trigger('click')

    expect(onClose).toBeCalledTimes(1)

    const closeIcon = h(IxIcon, { name: 'close' })
    await wrapper.setProps({ closeIcon })

    expect(modalWrapper.find('.ix-icon-close').exists()).toBe(true)
    expect(modalWrapper.find('.ix-icon-up').exists()).toBe(false)

    await modalWrapper.find('.ix-icon-close').trigger('click')

    expect(onClose).toBeCalledTimes(2)
  })

  test('closeIcon slot work', async () => {
    const wrapper = ModalMount({
      props: { closeIcon: 'up' },
      slots: {
        closeIcon: () => h(IxIcon, { name: 'close' }),
      },
    })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-icon-close').exists()).toBe(true)
    expect(modalWrapper.find('.ix-icon-up').exists()).toBe(false)
  })

  test('closeOnEsc work', async () => {
    const onUpdateVisible = vi.fn()
    const wrapper = ModalMount({ props: { closeOnEsc: false, 'onUpdate:visible': onUpdateVisible } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    await modalWrapper.trigger('keydown', { code: 'Escape' })

    expect(modalWrapper.find('.ix-modal').isVisible()).toBe(true)

    await wrapper.setProps({ closeOnEsc: true })

    await modalWrapper.trigger('keydown', { code: 'Escape' })

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('destroyOnHide work', async () => {
    const wrapper = ModalMount({ props: { visible: true, destroyOnHide: true } })
    let modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal').exists()).toBe(true)

    await wrapper.setProps({ visible: false })
    await flushPromises()

    expect(document.querySelector('.ix-modal-wrapper')).toBe(null)

    await wrapper.setProps({ visible: true })
    modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal').exists()).toBe(true)

    await wrapper.setProps({ destroyOnHide: false, visible: false })

    expect(modalWrapper.find('.ix-modal').exists()).toBe(true)
  })

  test('footer with boolean work', async () => {
    const wrapper = ModalMount({ props: { footer: false } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal-footer').exists()).toBe(false)

    await wrapper.setProps({ footer: true })

    expect(modalWrapper.find('.ix-modal-footer').exists()).toBe(true)
  })

  test('footer with buttons work', async () => {
    let footer: ModalButtonProps[] = [{ text: 'button1' }]
    const wrapper = ModalMount({ props: { footer } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.findAll('.ix-button').length).toBe(1)

    footer = [...footer, { text: 'button2' }]
    await wrapper.setProps({ footer })

    expect(modalWrapper.findAll('.ix-button').length).toBe(2)
  })

  test('footer with slot work', async () => {
    const footer: ModalButtonProps[] = [{ text: 'button1' }]
    const wrapper = ModalMount({
      props: { footer },
      slots: { footer: () => h(IxButton, {}, { default: () => 'button slot' }) },
    })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.findAll('.ix-button').length).toBe(1)
    expect(modalWrapper.find('.ix-button').text()).toBe('button slot')
  })

  test('header work', async () => {
    let header: string | HeaderProps = 'This is header'
    const wrapper = ModalMount({ props: { header } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-header').text()).toBe(header)

    header = 'This is header2'
    await wrapper.setProps({ header })

    expect(modalWrapper.find('.ix-header').text()).toBe(header)
    expect(modalWrapper.find('.ix-header').classes()).toContain('ix-header-md')

    header = { title: 'This is header2', size: 'lg' }
    await wrapper.setProps({ header })

    expect(modalWrapper.find('.ix-header').text()).toBe('This is header2')
    expect(modalWrapper.find('.ix-header').classes()).toContain('ix-header-lg')
  })

  test('header slot work', async () => {
    const header = 'This is header'
    const wrapper = ModalMount({
      props: { header },
      slots: {
        header: () => h(IxHeader, { title: 'this is header2' }),
      },
    })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-header').text()).toBe('this is header2')
  })

  test('icon work', async () => {
    const wrapper = ModalMount({ props: { type: 'confirm', icon: 'up' } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal-body-icon').find('.ix-icon-up').exists()).toBe(true)

    await wrapper.setProps({ icon: 'down' })

    expect(modalWrapper.find('.ix-modal-body-icon').find('.ix-icon-down').exists()).toBe(true)

    await wrapper.setProps({ icon: h(IxIcon, { name: 'up' }) })

    expect(modalWrapper.find('.ix-modal-body-icon').find('.ix-icon-up').exists()).toBe(true)

    await wrapper.setProps({ type: 'default' })
    expect(modalWrapper.find('.ix-modal-body-icon').exists()).toBe(false)
  })

  test('icon slot work', async () => {
    const wrapper = ModalMount({
      props: { type: 'confirm', icon: 'up' },
      slots: { icon: () => h(IxIcon, { name: 'down' }) },
    })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal-body-icon').find('.ix-icon-up').exists()).toBe(false)
    expect(modalWrapper.find('.ix-modal-body-icon').find('.ix-icon-down').exists()).toBe(true)
  })

  test('mask work', async () => {
    const wrapper = ModalMount({ props: { mask: false } })

    expect(isElementVisible(document.querySelector('.ix-mask'))).toBe(false)

    await wrapper.setProps({ mask: true })

    expect(isElementVisible(document.querySelector('.ix-mask'))).toBe(true)
  })

  test('maskClosable work', async () => {
    const onUpdateVisible = vi.fn()
    const wrapper = ModalMount({ props: { maskClosable: false, 'onUpdate:visible': onUpdateVisible } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    await modalWrapper.trigger('click')

    expect(onUpdateVisible).not.toBeCalled()

    await wrapper.setProps({ maskClosable: true })
    await modalWrapper.trigger('click')

    expect(onUpdateVisible).toBeCalledWith(false)
  })

  test('offset work', async () => {
    const wrapper = ModalMount({ props: { offset: 256 } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    const contentDom = modalWrapper.find('.ix-modal').element as HTMLElement

    expect(contentDom.style.top).toBe('256px')

    await wrapper.setProps({ offset: '30%' })

    expect(contentDom.style.top).toBe('30%')
  })

  test('title work', async () => {
    let title: VNode | string = 'This is title'
    const wrapper = ModalMount({ props: { type: 'confirm', title } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal-body-title').text()).toBe(title)

    title = 'This is title2'
    await wrapper.setProps({ title })

    expect(modalWrapper.find('.ix-modal-body-title').text()).toBe(title)

    title = createTextVNode('This is title3')
    await wrapper.setProps({ title })

    expect(modalWrapper.find('.ix-modal-body-title').text()).toBe('This is title3')

    await wrapper.setProps({ type: 'default' })
    expect(modalWrapper.find('.ix-modal-body-title').exists()).toBe(false)
  })

  test('title slot work', async () => {
    const wrapper = ModalMount({
      props: { type: 'confirm', title: 'This is title' },
      slots: { title: () => 'This is title2' },
    })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal-body-title').text()).toBe('This is title2')
  })

  test('type work', async () => {
    const wrapper = ModalMount({ props: { type: 'confirm' } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)

    expect(modalWrapper.find('.ix-modal-body-confirm').exists()).toBe(true)

    await wrapper.setProps({ type: 'info' })

    expect(modalWrapper.find('.ix-modal-body-info').exists()).toBe(true)

    await wrapper.setProps({ type: 'success' })

    expect(modalWrapper.find('.ix-modal-body-success').exists()).toBe(true)

    await wrapper.setProps({ type: 'warning' })

    expect(modalWrapper.find('.ix-modal-body-warning').exists()).toBe(true)

    await wrapper.setProps({ type: 'error' })

    expect(modalWrapper.find('.ix-modal-body-error').exists()).toBe(true)
  })

  test('width work', async () => {
    const wrapper = ModalMount({ props: { width: 400 } })
    const modalWrapper = wrapper.getComponent(ModalWrapper)
    const contentDom = modalWrapper.find('.ix-modal').element as HTMLElement

    expect(contentDom.style.width).toBe('400px')

    await wrapper.setProps({ width: '200px' })

    expect(contentDom.style.width).toBe('200px')

    await wrapper.setProps({ width: '20%' })

    expect(contentDom.style.width).toBe('20%')
  })

  test('zIndex work', async () => {
    const wrapper = ModalMount({ props: { zIndex: 1001 } })
    expect(document.querySelector('.ix-modal-wrapper')!.getAttribute('style')).toContain('z-index: 1001')

    await wrapper.setProps({ zIndex: 1002 })

    expect(document.querySelector('.ix-modal-wrapper')!.getAttribute('style')).toContain('z-index: 1002')
  })

  describe('Events', () => {
    test('onClose and onBeforeClose work', async () => {
      const onClose = vi.fn()
      const onBeforeClose = vi.fn()
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onClose, onBeforeClose, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.close()
      await flushPromises()

      expect(onClose).toBeCalled()
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onBeforeClose with result work', async () => {
      const onBeforeClose = vi.fn().mockImplementation((evt: unknown) => evt === 'close')
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onBeforeClose, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.close(1)
      await flushPromises()

      expect(onBeforeClose).toBeCalledWith(1)
      expect(onUpdateVisible).not.toBeCalled()

      wrapper.vm.close('close')
      await flushPromises()

      expect(onBeforeClose).toBeCalledWith('close')
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onBeforeClose with promise work', async () => {
      const onBeforeClose = vi.fn().mockImplementation((evt: unknown) => Promise.resolve(evt === 'close'))
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onBeforeClose, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.close(1)
      await flushPromises()

      expect(onBeforeClose).toBeCalledWith(1)
      expect(onUpdateVisible).not.toBeCalled()

      wrapper.vm.close('close')
      await flushPromises()

      expect(onBeforeClose).toBeCalledWith('close')
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onCancel work', async () => {
      const onCancel = vi.fn()
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onCancel, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.cancel()
      await flushPromises()

      expect(onCancel).toBeCalled()
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onCancel with result work', async () => {
      const onCancel = vi.fn().mockImplementation((evt: unknown) => evt === 'cancel')
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onCancel, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.cancel(1)
      await flushPromises()

      expect(onCancel).toBeCalledWith(1)
      expect(onUpdateVisible).not.toBeCalled()

      wrapper.vm.cancel('cancel')
      await flushPromises()

      expect(onCancel).toBeCalledWith('cancel')
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onCancel with promise work', async () => {
      const onCancel = vi.fn().mockImplementation((evt: unknown) => Promise.resolve(evt === 'cancel'))
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onCancel, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.cancel(1)
      await flushPromises()

      expect(onCancel).toBeCalledWith(1)
      expect(onUpdateVisible).not.toBeCalled()

      wrapper.vm.cancel('cancel')
      await flushPromises()

      expect(onCancel).toBeCalledWith('cancel')
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onOk work', async () => {
      const onOk = vi.fn()
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onOk, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.ok()
      await flushPromises()

      expect(onOk).toBeCalled()
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onOk with result work', async () => {
      const onOk = vi.fn().mockImplementation((evt: unknown) => evt === 'ok')
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onOk, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.ok(1)
      await flushPromises()

      expect(onOk).toBeCalledWith(1)
      expect(onUpdateVisible).not.toBeCalled()

      wrapper.vm.ok('ok')
      await flushPromises()

      expect(onOk).toBeCalledWith('ok')
      expect(onUpdateVisible).toBeCalledWith(false)
    })

    test('onOk with promise work', async () => {
      const onOk = vi.fn().mockImplementation((evt: unknown) => Promise.resolve(evt === 'ok'))
      const onUpdateVisible = vi.fn()
      const wrapper = ModalMount({ props: { onOk, 'onUpdate:visible': onUpdateVisible } })

      wrapper.vm.ok(1)
      await flushPromises()

      expect(onOk).toBeCalledWith(1)
      expect(onUpdateVisible).not.toBeCalled()

      wrapper.vm.ok('ok')
      await flushPromises()

      expect(onOk).toBeCalledWith('ok')
      expect(onUpdateVisible).toBeCalledWith(false)
    })
  })
})
