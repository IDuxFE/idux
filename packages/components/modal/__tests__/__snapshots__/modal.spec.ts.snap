// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal > render work 1`] = `
"<!--teleport start-->
<!--teleport end-->"
`;
